authors/apps/authentication/serializers.py

```diff
@@ -1,29 +1,138 @@
+    password = serializers.RegexField(
+        regex=r"^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d]{8,}$",
         max_length=128,
         min_length=8,
-        write_only=True
+        write_only=True,
+        required=True,
+        error_messages={
+            "required"  : "Password is required!",
+            "not_alphanum"   : "Password MUST have BOTH letters and numbers",
+            "too_short"   : "Password MUST be longer than eight(8) charcaters!",
+            "too_long"   : "Password is too long (max=128 characters)"
+        }
     )
-
     # The client should not be able to send a token along with a registration
     # request. Making `token` read-only handles that for us.
+    token = serializers.CharField(
+        read_only=True
+    )
 
     class Meta:
         model = User
         # List all of the fields that could possibly be included in a request
         # or response, including fields specified explicitly above.
-        fields = ['email', 'username', 'password']
+        fields = ['email', 'username', 'password', 'token']
+
+    def validate_username(self, data):
+        """
+        Validate the provided username
+        min_len=4
+        max_len=128
+        required=True
+        unique=True
+        """
+        candidate_name = data
+        
+        try:
+            if int(candidate_name):
+                raise serializers.ValidationError("Username cannot be numbers only!")
+        except ValueError:
+            pass
+            
+        if candidate_name == "":
+            raise serializers.ValidationError("Username is required!")
+        elif User.objects.filter(username=candidate_name):
+            raise serializers.ValidationError("Username already exists!")
+        elif len(candidate_name) < 4:
+            raise serializers.ValidationError("Username should be more than 4 charcaters")
+        elif len(candidate_name) > 128:
+            raise serializers.ValidationError("Username should not be longer than 128 charcaters")
+        return data
+
+    def validate_email(self, data):
+        """
+        Validate the provided email
+        required=True
+        unique=True
+        format: example@mail.com
+        """
+        candidate_email = data
+        if candidate_email == "":
+            raise serializers.ValidationError("Email is required!")
+        elif re.match(trial_email,candidate_email):
+            raise serializers.ValidationError("Invalid email! Hint: example@email.com")
+        elif re.match(trial_email_2, candidate_email):
+            raise serializers.ValidationError("Invalid email! Hint: example@email.com")
+        elif User.objects.filter(email=candidate_email):
+            raise serializers.ValidationError("User already exists. Please login!")
+        elif not re.match(email_expression, candidate_email):
+            raise serializers.ValidationError("The email is invalid! Hint: example@mail.com")
+        return data
+
+    def validate_password(self, data):
```

Comment:
Doesn't the `regex` value of the `password` field handle most of the checks below??

Reply:
I would advise you remove the validators on the password field and let this method handle most of the checks. The return statements are more descriptive and specific.

Reply:
I had included both the validators on the password field and the custom checks because I thought the validators on the field relate to how it is stored while the custom checks are meant for the payload. However, upon research, I have realised that the custom checks will serve both needs. As advised, I have retained the custom checks.